dask/array/core.py

```diff
@@ -1938,14 +1998,96 @@ def normalize_chunks(chunks, shape=None):
             raise ValueError("Input array has %d dimensions but the supplied "
                              "chunks has only %d dimensions" %
                              (len(shape), len(chunks)))
-        if not all(c == s or (math.isnan(c) and math.isnan(s))
+        if not all(c == s or (math.isnan(c) or math.isnan(s))
                    for c, s in zip(map(sum, chunks), shape)):
             raise ValueError("Chunks do not add up to shape. "
                              "Got chunks=%s, shape=%s" % (chunks, shape))
 
     return tuple(tuple(int(x) if not math.isnan(x) else x for x in c) for c in chunks)
 
 
+def auto_chunks(chunks, shape, limit, dtype, previous_chunks=None):
+    """ Determine automatic chunks
+
+    See also
+    --------
+    normalize_chunks: for full docstring and parameters
+    """
+    chunks = list(chunks)
+
+    autos = {i for i, c in enumerate(chunks) if c == 'auto'}
+    if not autos:
+        return tuple(chunks)
+
+    if limit is None:
+        limit = config.get('array.chunk-size')
+        if isinstance(limit, str):
+            limit = parse_bytes(limit)
+
+    if dtype is None:
+        raise TypeError("DType must be known for auto-chunking")
+
+    if dtype.hasobject:
+        raise NotImplementedError(
+            "Can not use auto rechunking with object dtype. "
+            "We are unable to estimate the size in bytes of object data")
+
+    limit = max(1, limit // dtype.itemsize)
+
+    largest_block = np.prod([cs if isinstance(cs, Number) else max(cs)
```

Comment:
`max(cs)` will break if any chunk sizes are NaN:
```
In [14]: max([1, np.nan])
Out[14]: 1

In [15]: max([np.nan, 1])
Out[15]: nan
```

`np.max()` handles this properly, though:
```
In [16]: np.max([1, np.nan])
Out[16]: nan

In [17]: np.max([np.nan, 1])
Out[17]: nan
```

Still, you might need special logic for NaN here -- or maybe just raise an error.

Reply:
We now explicitly raise an informative error if nans are used